src/v/raft/group_configuration.cc

```diff
@@ -260,194 +391,129 @@ void group_configuration::replace(
               auto it = std::find_if(
-                _brokers.begin(),
-                _brokers.end(),
+                _cfg._brokers.begin(),
+                _cfg._brokers.end(),
                 [&b](const model::broker& existing) {
                     return b.broker == existing;
                 });
 
-              return _current.contains(vnode(b.broker.id(), b.rev))
-                     && it != _brokers.end();
+              return _cfg._current.contains(vnode(b.broker.id(), b.rev))
+                     && it != _cfg._brokers.end();
           });
 
     // configurations are identical, do nothing
     if (brokers_are_equal) {
         return;
     }
 
-    _old = _current;
-    _current.learners.clear();
-    _current.voters.clear();
+    _cfg._old = _cfg._current;
+    _cfg._current.learners.clear();
+    _cfg._current.voters.clear();
 
     for (auto& br : brokers) {
         // check if broker is already a voter. voter will stay a voter
         auto v_it = std::find_if(
-          _old->voters.cbegin(), _old->voters.cend(), [&br](const vnode& rni) {
+          _cfg._old->voters.cbegin(),
+          _cfg._old->voters.cend(),
+          [&br](const vnode& rni) {
               return rni.id() == br.broker.id() && rni.revision() == br.rev;
           });
 
-        if (v_it != _old->voters.cend()) {
-            _current.voters.push_back(*v_it);
+        if (v_it != _cfg._old->voters.cend()) {
+            _cfg._current.voters.push_back(*v_it);
             continue;
         }
 
         // check if broker was a learner. learner will stay a learner
         auto l_it = std::find_if(
-          _old->learners.cbegin(),
-          _old->learners.cend(),
+          _cfg._old->learners.cbegin(),
+          _cfg._old->learners.cend(),
           [&br](const vnode& rni) {
               return rni.id() == br.broker.id() && rni.revision() == br.rev;
           });
 
-        if (l_it != _old->learners.cend()) {
-            _current.learners.push_back(*l_it);
+        if (l_it != _cfg._old->learners.cend()) {
+            _cfg._current.learners.push_back(*l_it);
             continue;
         }
 
         // new broker, use broker revision
-        _current.learners.emplace_back(br.broker.id(), br.rev);
+        _cfg._current.learners.emplace_back(br.broker.id(), br.rev);
     }
 
     // if both current and previous configurations are exactly the same, we do
     // not need to enter joint consensus
     if (
-      _current.voters == _old->voters && _current.learners == _old->learners) {
-        _old.reset();
+      _cfg._current.voters == _cfg._old->voters
+      && _cfg._current.learners == _cfg._old->learners) {
+        _cfg._old.reset();
     }
 
     for (auto& b : brokers) {
-        if (!contains_broker(b.broker.id())) {
-            _brokers.push_back(std::move(b.broker));
+        if (!_cfg.contains_broker(b.broker.id())) {
+            _cfg._brokers.push_back(std::move(b.broker));
```

Comment:
this all looks good. i keep thinking that if the group configuration were the virtual class and subclasses were the strategy specializations then all these _cfg prefixes might not be needed? either way, it's not a lot of change..

Reply:
i think that the serializtion/deserialization code wouldn't be very elegant